odl/ufunc_ops/ufunc_ops.py

```diff
@@ -0,0 +1,196 @@
+>>> op = {name}(space)
+>>> print(op({arg}))
+{result}
+"""
+
+
+class UfuncOperator(Operator):
+
+    """Base class for all ufunc operators."""
+
+
+def derivative_factory(name):
+    """Create derivative function for some ufuncs."""
+
+    if name == 'sin':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            point = self.domain.element(point)
+            return MultiplyOperator(point.ufunc.cos())
+        return derivative
+    elif name == 'cos':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            point = self.domain.element(point)
+            return MultiplyOperator(-point.ufunc.sin())
+        return derivative
+    elif name == 'tan':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            return MultiplyOperator(1 + self(point) ** 2)
+        return derivative
+    elif name == 'sqrt':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            return MultiplyOperator(0.5 / self(point))
+        return derivative
+    elif name == 'square':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            point = self.domain.element(point)
+            return MultiplyOperator(2.0 * point)
+        return derivative
+    elif name == 'log':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            point = self.domain.element(point)
+            return MultiplyOperator(1.0 / point)
+        return derivative
+    elif name == 'exp':
+        def derivative(self, point):
+            """Return the derivative operator."""
+            return MultiplyOperator(self(point))
+        return derivative
+    else:
+        # Fallback to default
+        return UfuncOperator.derivative
+
+
+def ufunc_class_factory(name, nargin, nargout, docstring):
+    """Create a UfuncOperator from a given specification."""
+
+    assert 0 <= nargin <= 2
+
+    def __init__(self, space):
+        """Initialize an instance."""
+        if not isinstance(space, LinearSpace):
+            raise TypeError('`space` {!r} not a `LinearSpace`'.format(space))
+
+        if _is_integer_only_ufunc(name) and not is_int_dtype(space.dtype):
+            raise ValueError('`space` {!r} not a `LinearSpace`'.format(space))
+
+        self.space = space
+
+        if nargin == 1:
+            domain = space
+        else:
+            domain = ProductSpace(space, nargin)
+
+        if nargout == 1:
+            range = space
+        else:
+            range = ProductSpace(space, nargout)
+
+        linear = name in LINEAR_UFUNCS
+        UfuncOperator.__init__(self, domain=domain, range=range, linear=linear)
+
+    def _call(self, x, out=None):
+        """return ``self(x)``."""
+        if out is None:
+            if nargin == 1:
+                return getattr(x.ufunc, name)()
+            elif nargin == 2:
```

Comment:
Why not `else`? The implementation will extend to that case, provided Numpy implements it this way, too. Alternatively, `nargin == 2` should be implemented explicitly and the `else` case raise a `NotImplementedError`. That will give us a chance to catch an unsupported upstream implementation.